open_rarity/rarity_ranker.py

```diff
@@ -20,35 +27,75 @@ def rank_tokens(*, token_id_to_scores: dict[str, float]) -> dict[str, int]:
 
         Parameters
         ----------
-        token_ids_to_score : dict[str, float]
-            A dictionary of token ids (this can be whatever id you want, as long
-            as its unique) to the rarity score.
+        collection : Collection
+            Collection object with populated tokens
+        scorer: Scorer
+            Scorer instance
 
         Returns
         -------
-        dict[str, int]
-            A dictionary of token ids to their ranks. More specificially, it is the
-            provided tokens ids to the rank it has within the set of possible
-            tokens defined by @token_id_to_scores.
+        list[TokenRarity]
+            list of TokenRarity objects with score, rank and token information
+        """
+
+        if (
+            collection is None
+            or collection.tokens is None
+            or len(collection.tokens) == 0
+        ):
+            return collection
+
+        tokens = collection.tokens
+        scores: list[float] = scorer.score_tokens(collection, tokens=tokens)
+
+        # fail ranking if dimension of scores doesn't match dimension of tokens
+        assert len(tokens) == len(scores)
+
+        token_rarity_list: list[TokenRarity] = []
+        # augment collection tokes with score information
+        for idx, token in enumerate(tokens):
+            token_rarity_list.append(
+                TokenRarity(token=token, score=scores[idx])
+            )
+
+        return RarityRanker.rank_tokens(token_rarity_list)
+
+    @staticmethod
+    def rank_tokens(tokens: list[TokenRarity]) -> list[TokenRarity]:
+        """Ranks a set of tokens Scores that are higher indicate a higher rarity,
+        and thus a lower rank.
+        Tokens with the same score will be assigned the same rank, e.g. we use RANK
+        (vs. DENSE_RANK).
+        Example: 1, 2, 2, 2, 5.
+        Scores are considered the same rank if they are within about 9 decimal digits
+        of each other.
+
+        Parameters
+        ----------
+        tokens : list[TokenRarity]
+            unordered list of tokens with rarity score information
+
+        Returns
+        -------
+        tokens: list[TokenRarity]
+            list of tokens with rarity scores sorted by rank
+
         """
-        sorted_token_ids = sorted(
-            token_id_to_scores.keys(),
-            key=lambda k: token_id_to_scores[k],
+        sorted_tokens_by_score: list[TokenRarity] = sorted(
```

Comment:
here too - i think should just be named
sorted_token_rarities (don't need by score, I think it's implied and clear enough)